src/service/navigation.js

```diff
@@ -738,6 +743,55 @@ export class Navigation {
       getMode().test && !this.isEmbed_ ? this.ampdoc.win.location.href : '';
     return this.parseUrl_(baseHref);
   }
+
+  /**
+   * Requests navigation through a Viewer to the given destination.
+   *
+   * This function only proceeds if:
+   * 1. The viewer supports the 'interceptNavigation' capability.
+   * 2. The contained AMP doc has 'opted in' via including the 'allow-navigation-interception'
+   * attribute on the <html> tag.
+   * 3. The viewer is trusted or from localhost.
+   *
+   * @param {string} url A URL.
+   * @param {string} requestedBy Informational string about the entity that
+   *     requested the navigation.
+   * @return {boolean} Returns true if navigation message was sent to viewer.
+   *     Otherwise, returns false.
+   */
+  navigateToUrlWithViewer(url, requestedBy) {
+    const viewerHasCapability = this.viewer_.hasCapability(
+      'interceptNavigation'
+    );
+    const docOptedIn = this.ampdoc
+      .getRootNode()
+      .documentElement.hasAttribute('allow-navigation-interception');
+
+    if (!viewerHasCapability || !docOptedIn) {
+      console.log('Did not navigate with viewer');
+      return false;
+    }
+
+    return Promise.all([
+      this.viewer_.isTrustedViewer(),
+      this.viewer_.getViewerOrigin(),
+    ]).then((values) => {
```

Comment:
We're going to have some asynchronicity issues here. We can only tell if the viewer is trusted async, but we need to call `preventDefault` sync (if needed). /cc @choumx

Reply:
Can we just evaluate `isTrustedViewer` in the constructor, similar to `this.appendExtraParams_`? Seems like that should probably resolve before the user has the chance to click anyway

Reply:
Good catch. There was a similar case "appendExtraParams" that put the async function as part of the constructor, checking the result synchronously when needed, so I borrowed from that in the new version. Now, we run the async calls as part of the constructor and store the results to be used synchronously. There is a case where the promises have not resolved by the time a user has clicked a link, but it seems unlikely for this to be the case. Let me know what you think.